### PR TITLE
[Clang] Update missing varargs arg extension warnings

### LINKED ISSUE #84495: Zero variable arguments should be allowed in C23
Clang doesn't allow zero variable arguments in macro invocations of macros with named arguments and then variable arguments. Here is an example program (https://godbolt.org/z/xbacsGqbM):
```c
#include<stdarg.h>
void foo(...){
    va_list v;
    va_start(v);
    va_arg(v,int(*)[5]);
    va_end(v);
}
int main(void){
    foo(&(int[]){1,2,3,4,5});
}
```
Clang complains here about not having a comma in `va_start` and wants `va_start(v,)`. This restriction was removed in C23 so this code should be allowed without any complaint.

### PR DESCRIPTION
**Pull request by Sirraide**
This updates a few warnings that were diagnosing no arguments for a `...` variadic macro parameter as a GNU extension when it actually is a C++20/C23 extension now.

This fixes #84495.

### CODE REVIEW

File: clang/lib/Lex/PPMacroExpansion.cpp [hunk 1]
```diff
@@ -993,11 +993,18 @@ MacroArgs *Preprocessor::ReadMacroCallArgumentList(Token &MacroName,
       // If the macro contains the comma pasting extension, the diagnostic
       // is suppressed; we know we'll get another diagnostic later.
       if (!MI->hasCommaPasting()) {
-        // C++20 allows this construct, but standards before C++20 and all C
-        // standards do not allow the construct (we allow it as an extension).
-        Diag(Tok, getLangOpts().CPlusPlus20
-                      ? diag::warn_cxx17_compat_missing_varargs_arg
-                      : diag::ext_missing_varargs_arg);
+        // C++20 and C23 allow this construct, but standards before that
+        // do not (we allow it as an extension).
```
Comment (shafik) [inline clang/lib/Lex/PPMacroExpansion.cpp:997]: CC @AaronBallman should we add a standard quote like we normally do for C++? It looks like it should be 6.10.5 p12. It took me a but to remember in C they are referred to as *variable arguments* and not variadic like in C++.

**Review by shafik — COMMENTED** [review]
Just a nit here.

File: clang/lib/Lex/PPMacroExpansion.cpp (hunk 1 above)
Comment (AaronBallman) [reply clang/lib/Lex/PPMacroExpansion.cpp:997]: I think it's worth adding the standards citation just the same, even if we don't copy the text from the standard. e.g.,
```
// C++20 [cpp.replace]p15, C23 6.10.5p12
```

**Review by AaronBallman — APPROVED** [review]
LGTM aside from a comment request

File: clang/lib/Lex/PPMacroExpansion.cpp (hunk 1 above)
Comment (Sirraide) [reply clang/lib/Lex/PPMacroExpansion.cpp:997]: Sure, will do.

**Comment by zibi2** [comment]
@Sirraide, I noticed the following build issue when new compiler with this change (bootstrap compiler)  is being used to build one of the files for `clangd`:

```
.../clang-tools-extra/clangd/unittests/FindTargetTests.cpp:430:29: error: passing no argument for the '...' parameter of a variadic macro is a C++20 extension [-Werror,-Wc++20-extensions]
  430 |   EXPECT_DECLS("AutoTypeLoc");
      |                             ^
.../clang-tools-extra/clangd/unittests/FindTargetTests.cpp:98:9: note: macro 'EXPECT_DECLS' defined here
   98 | #define EXPECT_DECLS(NodeType, ...)                                            \
      |         ^
```

Do you have suggestion how to fix it or even better provie the fix?
Thank you,
Zibi

**Comment by Sirraide** [comment]
> @Sirraide, I noticed the following build issue when new compiler with this change (bootstrap compiler) is being used to build one of the files for `clangd`:
> 
> ```
> .../clang-tools-extra/clangd/unittests/FindTargetTests.cpp:430:29: error: passing no argument for the '...' parameter of a variadic macro is a C++20 extension [-Werror,-Wc++20-extensions]
>   430 |   EXPECT_DECLS("AutoTypeLoc");
>       |                             ^
> .../clang-tools-extra/clangd/unittests/FindTargetTests.cpp:98:9: note: macro 'EXPECT_DECLS' defined here
>    98 | #define EXPECT_DECLS(NodeType, ...)                                            \
>       |         ^
> ```
> 
> Do you have suggestion how to fix it or even better provie the fix? Thank you, Zibi

I mean, from what I can tell, this isn’t a Clang issue but rather just a result of this code here not being standards-compliant C++17. Before C++20, afaik the ‘fix’ for this is to have two different macros: one variadic, one not variadic. It’s not pretty, but that’s why that got changed in C++20.

**Comment by Sirraide** [comment]
Actually, a trailing comma in the macro invocation might work too. It’s been a while since I last had to deal w/ this.

**Comment by zibi2** [comment]
I agree the code is not C++17 compliant. An alternative fix would be to compile this file with `-Wno-c++20-extensions`.

**Comment by zibi2** [comment]
> Actually, a trailing comma in the macro invocation might work too. It’s been a while since I last had to deal w/ this.

Yes, it does and I think it's cleaner than my proposal. Thank you.

**Pull request merged**